### PR TITLE
 feat: Add seed job for monorepo layout.

### PR DESCRIPTION
**Pull request by Wayonb**
Add the seed job which can be used to automatically create the other package jobs.
The seed job can be configured to update the package jobs if changes are made to the buildConfiguration.yaml file.
The ``buildConfiguration.yaml`` has layout of each package and their dependencies.

### CODE REVIEW

File: jenkins/shared-library/vars/createMonorepoMultibranchJobs.groovy [hunk 1]
```diff
@@ -40,14 +40,13 @@ void generatePackageMultibranchJobs(Object buildConfiguration, Map jobConfigurat
...
 		String pipelineName = build.path.tokenize(pathSeparator).last()
-		jobConfiguration.with {
-			jobName = Paths.get(jobConfiguration.fullBranchFolder.toString()).resolve(pipelineName).toString()
-			jenkinsfilePath = Paths.get(build.path).resolve('Jenkinsfile').toString()
-			packageIncludePaths = addPathAndDependsOnFolder(build).join('\n')
-			packageExcludePaths = '**/*'
-			path = build.path
-			packageFolder = Paths.get(jobConfiguration.repositoryName.toString()).resolve(build.path).toString()
-		}
+		jobConfiguration.jobName = Paths.get(jobConfiguration.fullBranchFolder.toString()).resolve(pipelineName).toString()
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/createMonorepoMultibranchJobs.groovy:43]: double check `with` behavior as discussed in Discord. at minimum, you'll need need to suppress warning(s).

File: seed-job/buildConfiguration.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+builds:
+  - name: Rest-Gateway
+    path: client/rest
```
Comment (Jaguar0625) [inline seed-job/buildConfiguration.yaml:3]: can you separate each entry with newline?

File: seed-job/buildConfiguration.yaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    dependsOn:
+      - catbuffer/parser
+  - name: Sdk-Javascript
+    path: sdk/javascript
+    dependsOn:
+      - catbuffer/parser
+  - name: Jenkins
+    path: jenkins
+  - name: Linters
+    path: linters
```
Comment (Jaguar0625) [inline seed-job/buildConfiguration.yaml:17]: newline at end of file

File: seed-job/buildConfiguration.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+builds:
+  - name: Rest-Gateway
```
Comment (Jaguar0625) [inline seed-job/buildConfiguration.yaml:2]: do you need - or can you use space?

File: seed-job/buildConfiguration.yaml (hunk 3 above)
Comment (Jaguar0625) [inline seed-job/buildConfiguration.yaml:17]: should everything depend on linters and/or jenkins? 🤔

File: seed-job/Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,5 @@
+monorepoSeedPipeline {
+  platform = ['ubuntu']
+  organizationName = 'Symbol'
+  gitHubId = 'Symbol-Github-app'
```
Comment (Jaguar0625) [inline seed-job/Jenkinsfile:4]: check if needed - see #46 git checkout removal

File: seed-job/buildConfiguration.yaml (hunk 4 above)
Comment (gimre-xymcity) [reply seed-job/buildConfiguration.yaml:2]: I guess need, as those are "array" entries, right?

File: jenkins/shared-library/vars/createMonorepoMultibranchJobs.groovy (hunk 1 above)
Comment (Wayonb) [reply jenkins/shared-library/vars/createMonorepoMultibranchJobs.groovy:43]: I am incline to suppress unless you have any other ideas.
I took your example code and ran it in Jenkins script console and it works.  In the pipeline, there might be a bug or I am missing something.   The values that gets assign in the ``with`` statement dont exist after the ``with`` block.

```
		println "Hello World!"

		def jobConfiguration = [:]
		jobConfiguration.alpha = 'hello'
		jobConfiguration.zeta = 'dear'

		jobConfiguration.with {
			println('start with')
			beta = 'foo'
			zeta = 'bar'
			println("end with beta: ${beta}")
		};

		println("alpha: ${jobConfiguration.alpha}")
		println("beta: ${jobConfiguration.beta}")
		println("zeta: ${jobConfiguration.zeta}")
	``` 

The out in the pipeline is below
```
[Pipeline] echo
10:43:16  Hello World!
[Pipeline] echo
10:43:16  start with
[Pipeline] echo
10:43:16  end with beta: foo
[Pipeline] echo
10:43:16  alpha: hello
[Pipeline] echo
10:43:16  beta: null
[Pipeline] echo
10:43:16  zeta: dear
```

Comment (Jaguar0625) [reply jenkins/shared-library/vars/createMonorepoMultibranchJobs.groovy:43]: can you try something like this?
```
println "Hello World!"

def jobConfiguration = [:]
jobConfiguration.alpha = 'hello'
jobConfiguration.zeta = 'dear'

def newJobConfigurationProperties = [:]
newJobConfigurationProperties.with{
	println('start with')
	beta = 'foo'
	zeta = 'bar'
	println("end with beta: ${beta}")
};

jobConfiguration += newJobConfigurationProperties

println("alpha: ${jobConfiguration.alpha}")
println("beta: ${jobConfiguration.beta}")
println("zeta: ${jobConfiguration.zeta}")
```

Comment (Wayonb) [reply jenkins/shared-library/vars/createMonorepoMultibranchJobs.groovy:43]: I tried several different variation but still no luck
```
[Pipeline] echo
11:31:31  Hello World!
[Pipeline] echo
11:31:31  start with
[Pipeline] echo
11:31:31  end with beta: foo
[Pipeline] echo
11:31:31  newJobConfigurationProperties: [:]
[Pipeline] echo
11:31:31  alpha: hello
[Pipeline] echo
11:31:31  beta: null
[Pipeline] echo
11:31:31  zeta: dear
```

File: seed-job/Jenkinsfile (hunk 5 above)
Comment (Wayonb) [reply seed-job/Jenkinsfile:4]: Only needed for private repo or if we still plan to automatic release where the Jenkins needs to update the Github

File: seed-job/buildConfiguration.yaml (hunk 4 above)
Comment (Wayonb) [reply seed-job/buildConfiguration.yaml:2]: yes, its an array.

File: seed-job/buildConfiguration.yaml (hunk 3 above)
Comment (Wayonb) [reply seed-job/buildConfiguration.yaml:17]: maybe after we get to a stable state?

File: seed-job/buildConfiguration.yaml (hunk 4 above)
Comment (Jaguar0625) [reply seed-job/buildConfiguration.yaml:2]: i meant `Rest-Gateway` vs `Rest Gateway`

File: jenkins/shared-library/vars/createMonorepoMultibranchJobs.groovy (hunk 1 above)
Comment (Jaguar0625) [reply jenkins/shared-library/vars/createMonorepoMultibranchJobs.groovy:44]: :( can you try replacing `with` with `identity` ?

Comment (Jaguar0625) [reply jenkins/shared-library/vars/createMonorepoMultibranchJobs.groovy:44]: seems like known issue: https://stackoverflow.com/questions/57561261/why-does-with-method-doesnt-work-in-jenkins-pipeline
😢

File: jenkins/shared-library/vars/createMonorepoMultibranchJobs.groovy [hunk 6]
```diff
@@ -60,13 +60,13 @@ void generateRepoMultibranchJob(Object buildConfiguration, Map jobConfiguration)
...
-		jobName = Paths.get(jobConfiguration.fullBranchFolder).resolve(pipelineName).toString()
-		jenkinsfilePath = 'Jenkinsfile'
-		packageIncludePaths = ''
-		packageFolder = Paths.get(jobConfiguration.repositoryName).resolve(pipelineName).toString()
-	}
+	jobConfiguration.jobName = Paths.get(jobConfiguration.fullBranchFolder).resolve(pipelineName).toString()
+	jobConfiguration.displayName = pipelineName
+	jobConfiguration.jenkinsfilePath = 'Jenkinsfile'
+	jobConfiguration.packageIncludePaths = ''
+	jobConfiguration.packageFolder = Paths.get(jobConfiguration.repositoryName).resolve(pipelineName).toString()
```
Comment (Jaguar0625) [inline jenkins/shared-library/vars/createMonorepoMultibranchJobs.groovy:69]: unless there is good reason, assign properties in same order here and above

File: seed-job/Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,5 @@
+monorepoSeedPipeline {
```
Comment (Jaguar0625) [inline seed-job/Jenkinsfile:1]: tabify

**Review by Jaguar0625 — APPROVED** [review]

**Pull request merged**